Applications/SimpleShot/layers/centering.cpp

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   centering.cpp
+ * @date   08 Jan 2021
+ * @brief  This file contains the simple centering layer which has hardcoded
+ * mean feature vectors from given combinations
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <fstream>
+#include <iostream>
+#include <regex>
+#include <sstream>
+
+#include <nntrainer_error.h>
+#include <tensor.h>
+#include <tensor_dim.h>
+
+#include <centering.h>
+#include <simpleshot_utils.h>
+
+namespace simpleshot {
+namespace layers {
+
+const std::string CenteringLayer::type = "centering";
+
+CenteringLayer::CenteringLayer(const std::string &feature_path_) :
+  Layer(),
+  feature_path(feature_path_) {}
+
+int CenteringLayer::setProperty(std::vector<std::string> values) {
+  util::Entry e;
+
+  std::vector<std::string> unhandled_values;
+
+  for (auto &val : values) {
+    try {
+      e = util::getKeyValue(val);
+    } catch (std::invalid_argument &e) {
+      std::cerr << e.what() << std::endl;
+      return -1;
+    }
+
+    if (e.key == "feature_path") {
+      feature_path = e.value;
+    } else {
+      unhandled_values.push_back(val);
+    }
+  }
+
+  return nntrainer::Layer::setProperty(unhandled_values);
+}
+
+int CenteringLayer::initialize(nntrainer::Manager &manager) {
+  output_dim[0] = input_dim[0];
+  mean_feature_vector = nntrainer::Tensor(input_dim[0]);
+  std::ifstream file(feature_path, std::ios::in | std::ios::binary);
+  mean_feature_vector.read(file);
+
+  return ML_ERROR_NONE;
+}
+
+void CenteringLayer::forwarding(bool training) {
+  net_input[0]->getVariableRef().add(mean_feature_vector,
+                                     net_hidden[0]->getVariableRef(), -1);
+}
+
+void CenteringLayer::calcDerivative() {
+  // do nothing
```

Comment:
This should throw.

Reply:
Currently this function is used by `neuralnet::train()`. It it okay to throw?

Reply:
If setTrainable(false) it still calls this function. Am I right??

Reply:
If setTrainable() is false, and all the layers behind this are also not trainable, then this won't be called.
Otherwise, it will be called even with setTrainable() to false.

Reply:
getTrainable() is overriden because it always returns false